saleor/payment/tasks.py

```diff
@@ -30,131 +31,143 @@ def checkouts_with_funds_to_release():
         - settings.CHECKOUT_TTL_BEFORE_RELEASING_FUNDS
     )
 
-    return Checkout.objects.using(settings.DATABASE_CONNECTION_REPLICA_NAME).filter(
-        Q(
-            automatically_refundable=True,
-            last_change__lt=expired_checkouts_time,
-            last_transaction_modified_at__lt=expired_checkouts_time,
-        )
-        & (
-            ~Q(authorize_status=CheckoutAuthorizeStatus.NONE)
-            | ~Q(charge_status=CheckoutChargeStatus.NONE)
-        )
+    checkouts = Checkout.objects.using(
+        settings.DATABASE_CONNECTION_REPLICA_NAME
+    ).filter(
+        automatically_refundable=True,
+        last_change__lt=expired_checkouts_time,
+        last_transaction_modified_at__lt=expired_checkouts_time,
+        authorize_status__in=[
+            CheckoutAuthorizeStatus.PARTIAL,
+            CheckoutAuthorizeStatus.FULL,
+        ],
     )
 
+    # Fetch transactions for checkouts that are ready to release funds.
+    transactions = (
+        TransactionItem.objects.using(settings.DATABASE_CONNECTION_REPLICA_NAME).filter(
+            Q(
+                Exists(checkouts.filter(pk=OuterRef("checkout_id"))),
+                order_id=None,
+                last_refund_success=True,
+            )
+            & (Q(authorized_value__gt=0) | Q(charged_value__gt=0))
+        )
+    ).order_by("created_at")
+    return transactions
+
 
 @app.task
 @allow_writer()
 def transaction_release_funds_for_checkout_task():
-    CHECKOUT_BATCH_SIZE = int(settings.CHECKOUT_BATCH_FOR_RELEASING_FUNDS)
     TRANSACTION_BATCH_SIZE = int(settings.TRANSACTION_BATCH_FOR_RELEASING_FUNDS)
 
-    # Fetch checkouts that are ready to release funds
-    checkouts = checkouts_with_funds_to_release().order_by("last_change")
-    checkouts_data = list(
-        checkouts.values_list("pk", "channel_id")[:CHECKOUT_BATCH_SIZE]
+    # Fetch transactions that are ready to release funds
+    transactions = transactions_to_release_funds().order_by("modified_at")
+    transactions_data = list(
+        transactions.values_list("pk", "checkout_id")[:TRANSACTION_BATCH_SIZE]
+    )
+    transaction_pks = [pk for pk, _checkout_id in transactions_data]
+    checkout_ids = [checkout_id for _, checkout_id in transactions_data]
+
+    checkouts_data = Checkout.objects.filter(pk__in=checkout_ids).values_list(
+        "pk", "channel_id"
     )
-    checkout_pks = [pk for pk, _ in checkouts_data]
     checkout_channel_ids = [channel_id for _, channel_id in checkouts_data]
```

Comment:
It is only a detail, as we don't have a lot of `checkouts_data` fetched at once, but to keep it cleaner, I would suggest adding `set()` here. No need to add in `filter()` duplicated values.